In [2]:
! pip install numpy

1. **Difference in terms of memory usage**

        List:  stores references to objects
               Can store mixed type elements	
               Memory Usage is more (stores type + reference info)

        Array: stores elements in contiguous memory
               only stores same type elements
               Memory Usage is less(compact, fixed type)

   **Difference in terms of performance**

        List:  require loops for operations
               can do basic arithmetic calculation only

        Array: use vectorized operations
               faster for mathematical operations.                


In [3]:
import numpy as np

# Python Lists
l1 = [1, 2, 3, 4, 5]
l2 = [10, 20, 30, 40, 50]

re = []

for i in range(len(l1)):
    re.append(l1[i] * l2[i])

print("List Multiplication:", re)


# NumPy Arrays
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])
r = a * b
print("NumPy Multiplication:", r)

List Multiplication: [10, 40, 90, 160, 250]
NumPy Multiplication: [ 10  40  90 160 250]


2. Broadcasting means NumPy automatically stretches the smaller array to match the bigger array so mathematical operations can happen.It does this without actually copying the   data.
 
   A is shape (3,3).

   B is shape (3,).

   NumPy treats B as (1,3).

   It repeats B across rows to match (3,3).
   
   Then performs element-wise addition.

Example: 

        1.  [1  2  3
               5  6  7]

        2.  wants to add row 
            [8 9 0]

         3. so second row will be 
            [8 9 0
             8 9 0]   

In [4]:
import numpy as np

a = np.array([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])

b = np.array([10, 20, 30])

r = a + b
print(r)


[[11 22 33]
 [14 25 36]
 [17 28 39]]


3. Missing values are incomplete or unavailable data.

Representation:

       NaN (Not a Number)

      None

      pd.NA

 **df.isnull()** Shows True where value is missing

 **df.mean()** Calculates average of each column

 **fillna()** Replaces missing values with that average  

In [5]:
import pandas as pd
import numpy as np

# Create DataFrame with missing values
df = pd.DataFrame({
    "age": [20, 25, np.nan, 30, np.nan],
    "salary": [50000, np.nan, 60000, 55000, 52000]
})

print("Original DataFrame:")
print(df)

# Detect missing values
print("missing values: ")
print(df.isnull())

# Replace missing values with column mean
df = df.fillna(df.mean())

print("after replacing missing vlues with mean: ")
print(df)


Original DataFrame:
    age   salary
0  20.0  50000.0
1  25.0      NaN
2   NaN  60000.0
3  30.0  55000.0
4   NaN  52000.0
missing values: 
     age  salary
0  False   False
1  False    True
2   True   False
3  False   False
4   True   False
after replacing missing vlues with mean: 
    age   salary
0  20.0  50000.0
1  25.0  54250.0
2  25.0  60000.0
3  30.0  55000.0
4  25.0  52000.0


4. Boolean indexing is a method used to filter data using True/False (boolean) conditions.When you apply a condition to a NumPy array or Pandas DataFrame, it returns a boolean mask (True/False values). It is one of the most powerful features for data filtering.

True if row is selected and False if row is removed


The condition (like Age > 25) is applied to every element in the column. This produces a boolean mask — a list/array of True and False values.

Each row gets **True** if it satisfies the condition and **False** if it does not.

When multiple conditions are used, logical operators (AND, OR) combine these boolean masks element-by-element.

Internally, Pandas/NumPy use optimized C-based vectorized operations to keep rows where the mask is True and discard rows where the mask is False

In [9]:
import pandas as pd

data = {
    "Name": ["Aman", "Riya", "John", "Priya", "Rahul"],
    "Age": [22, 25, 30, 28, 24],
    "Department": ["IT", "HR", "IT", "Finance", "HR"],
    "Salary": [40000, 35000, 60000, 50000, 45000],
    "Experience": [1, 3, 5, 4, 2]
}

df = pd.DataFrame(data)

filtered_df = df[(df["Age"] > 25) & (df["Salary"] > 45000)]
print(filtered_df)

    Name  Age Department  Salary  Experience
2   John   30         IT   60000           5
3  Priya   28    Finance   50000           4


5. The groupby() function is used to Split -> Apply -> Combine data.

It:

Splits data into groups based on category

Applies aggregation function (mean, sum, count, etc.)

Combines results into a new DataFrame/Series

In [8]:
data = {
    "Employee": ["Aman", "Riya", "John", "Priya", "Rahul", "Sneha"],
    "Department": ["IT", "HR", "IT", "Finance", "HR", "IT"],
    "Salary": [40000, 35000, 60000, 50000, 45000, 55000]
}

df = pd.DataFrame(data)

avg_salary = df.groupby("Department")["Salary"].mean()
print(avg_salary)

Department
Finance    50000.000000
HR         40000.000000
IT         51666.666667
Name: Salary, dtype: float64


**Step 1: Split**

Pandas groups rows by unique Department values:

IT → (40000, 60000, 55000)

HR → (35000, 45000)

Finance → (50000)

**Step 2: Apply**

It applies mean() to each group:

IT → (40000 + 60000 + 55000) / 3 = 51666.7

HR → (35000 + 45000) / 2 = 40000

Finance → 50000

**Step 3: Combine**

Pandas returns:

A Series

Indexed by Department

Containing average salary